In [ ]:
import scipy.io
import pandas as pd
import warnings
warnings.filterwarnings("ignore")
file_path = 'balanced_coast_dictionary.mat' # fr others change middle word to corrosponding dic vvv

mat_data = scipy.io.loadmat(file_path)
matrix_data = mat_data['dic_coast'] # for others use dic_land, dic_ocean, dic_coast, dic_seaIce (orig is dic_snow)

# Convert the matrix data to a Pandas DataFrame
df = pd.DataFrame(matrix_data) # load dictionary for different surface type and store it into differet dataframe

In [ ]:
rows_to_keep = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 14, 29] # I have put the input features in the rows
#and only the rows with given rows index will be used for our model
df_req = df.iloc[rows_to_keep]
df_req

0           1           2           3           4           5      \
0   188.050003  240.210007  162.130005  194.509995  163.110001  187.199997   
1   117.709999  191.199997   90.650002  132.199997   88.269997  127.400002   
2   205.350006  253.660004  179.630005  196.100006  178.869995  199.130005   
3   144.979996  217.059998  108.889999  132.740005  105.320000  137.529999   
4   241.550003  265.769989  199.419998  210.410004  193.630005  213.679993   
5   223.729996  266.920013  205.470001  214.899994  204.240005  220.029999   
6   167.949997  238.820007  137.979996  151.779999  136.949997  162.750000   
7   274.649994  284.769989  243.889999  251.199997  242.759995  247.479996   
8   255.800003  276.010010  195.800003  203.949997  194.389999  208.600006   
9   285.190002  280.940002  265.670013  278.179993  266.190002  263.880005   
10  284.230011  280.350006  258.429993  269.459991  252.500000  257.279999   
11  266.739990  256.350006  249.289993  264.339996  262.059998  252.240005   
12  276.440002  269.510010  263.380005  276.950012  268.739990  260.100006   
14    0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
29         NaN    0.000000         NaN         NaN         NaN    0.000000   

         6           7           8           9      ...       49284  \
0   268.260010  164.869995  226.089996  225.919998  ...  227.259995   
1   241.259995   88.860001  191.990005  175.179993  ...  196.589996   
2   254.520004  182.910004  247.759995  245.289993  ...  242.279999   
3   230.850006  112.610001  220.130005  209.429993  ...  218.710007   
4   262.970001  207.779999  251.479996  261.549988  ...  251.110001   
5   259.769989  206.679993  258.519989  255.520004  ...  255.399994   
6   238.309998  141.589996  239.529999  223.800003  ...  237.679993   
7   278.500000  251.179993  267.890015  284.829987  ...  264.170013   
8   273.010010  210.779999  257.250000  277.200012  ...  258.899994   
9   281.130005  272.980011  270.260010  288.100006  ...  264.029999   
10  280.339996  270.329987  268.959991  285.989990  ...  261.570007   
11  261.959991  254.240005  255.899994  265.769989  ...  248.940002   
12  273.329987  265.109985  265.750000  277.299988  ...  257.109985   
14    0.000000    0.000000    0.000000    0.000000  ...    3.000000   
29         NaN         NaN         NaN         NaN  ...         NaN   

         49285       49286       49287       49288       49289       49290  \
0   225.960007  225.960007  225.960007  225.960007  225.960007  225.960007   
1   192.759995  192.759995  192.759995  192.759995  192.759995  192.759995   
2   239.479996  239.479996  239.479996  239.479996  239.479996  239.479996   
3   213.809998  213.809998  213.809998  213.809998  213.809998  213.809998   
4   249.360001  249.360001  249.360001  249.360001  249.360001  249.360001   
5   252.800003  252.800003  252.800003  252.800003  252.800003  252.800003   
6   233.610001  233.610001  233.610001  233.610001  233.610001  233.610001   
7   263.820007  263.820007  263.820007  263.820007  263.820007  263.820007   
8   257.260010  257.260010  257.260010  257.260010  257.260010  257.260010   
9   264.630005  264.630005  264.630005  264.630005  264.630005  264.630005   
10  262.859985  262.859985  262.859985  262.859985  262.859985  262.859985   
11  247.869995  247.869995  247.869995  247.869995  247.869995  247.869995   
12  258.279999  258.279999  258.279999  258.279999  258.279999  258.279999   
14    3.000000    3.000000    3.000000    3.000000    3.000000    3.000000   
29         NaN         NaN         NaN         NaN         NaN         NaN   

         49291       49292       49293  
0   225.130005  225.130005  225.130005  
1   192.880005  192.880005  192.880005  
2   239.350006  239.350006  239.350006  
3   211.160004  211.160004  211.160004  
4   249.160004  249.160004  249.160004  
5   251.610001  251.610001  251.610001  
6   231.279999  231.279999  231.279999  
7   263.390015  263.390015  263.390015  
8   2

In [ ]:
df_req.iloc[13].unique() # This rows give information about precipitation phase. 0 = no_precip, 3= rain, 5= snow

array([0., 5., 3.])

In [ ]:
df_req.iloc[13] = df_req.iloc[13].replace({3.0: 1, 5.0: 2}) # Modifying the index so 1 = rain, 2=snow
# we are providing the name to the input features. They are the information measured by the differnt frequencies that
# helps to determine the precipitation phase
variables = ['10v_BT', '10h_BT', '18v_BT', '18h_BT', '23v_BT', '36v_BT', '36h_BT', '89v_BT', '89h_BT', '166v_BT', '166h_BT', '183-3_BT','183-7_BT', 'precip_flag','pr_rate_nearSurf']
df_req.loc[:, 'variable'] = variables
df_req.set_index('variable', inplace=True)
# If you find missing data, delete the entire column. If the data of one input feature is missing, data of other features at
#that instant is of no use
df_req = df_req.dropna(axis=1)
#I am transposing it for the proper visualization
df_tran = df_req.T
df_tran

variable      10v_BT      10h_BT      18v_BT      18h_BT      23v_BT  \
1         240.210007  191.199997  253.660004  217.059998  265.769989   
5         187.199997  127.400002  199.130005  137.529999  213.679993   
11        231.029999  198.130005  257.890015  241.520004  262.829987   
14        170.490005  103.790001  177.539993  104.120003  194.020004   
15        216.929993  173.729996  216.089996  171.710007  224.440002   
...              ...         ...         ...         ...         ...   
49205     216.149994  176.580002  231.740005  203.210007  245.229996   
49206     216.149994  176.580002  231.740005  203.210007  245.229996   
49207     216.149994  176.580002  231.740005  203.210007  245.229996   
49208     216.149994  176.580002  231.740005  203.210007  245.229996   
49209     226.130005  191.809998  225.000000  185.960007  234.550003   

variable      36v_BT      36h_BT      89v_BT      89h_BT     166v_BT  \
1         266.920013  238.820007  284.769989  276.010010  280.940002   
5         220.029999  162.750000  247.479996  208.600006  263.880005   
11        265.489990  256.399994  270.640015  267.160004  267.619995   
14        204.639999  133.679993  241.990005  186.929993  261.429993   
15        234.320007  196.500000  252.089996  217.289993  264.660004   
...              ...         ...         ...         ...         ...   
49205     247.309998  226.410004  262.220001  255.529999  257.619995   
49206     247.309998  226.410004  262.220001  255.529999  257.619995   
49207     247.309998  226.410004  262.220001  255.529999  257.619995   
49208     247.309998  226.410004  262.220001  255.529999  257.619995   
49209     236.789993  198.089996  255.610001  220.740005  272.000000   

variable     166h_BT    183-3_BT    183-7_BT  precip_flag  pr_rate_nearSurf  
1         280.350006  256.350006  269.510010          0.0          0.000000  
5         257.279999  252.240005  260.100006          0.0          0.000000  
11        267.420013  251.380005  259.920013          0.0          0.000000  
14        248.110001  252.770004  262.549988          0.0          0.000000  
15        249.639999  252.800003  262.679993          0.0          0.000000  
...              ...         ...         ...          ...               ...  
49205     254.039993  246.880005  252.589996          1.0          0.431430  
49206     254.039993  246.880005  252.589996          1.0          0.431430  
49207     254.039993  246.880005  252.589996          1.0          0.431430  
49208     254.039993  246.880005  252.589996          1.0          0.685017  
49209     263.160004  257.779999  270.130005          1.0          0.000000  

[15590 rows x 15 columns]

Now, you can start training the model. Create the datafrmae like this for the different surface type and evaluate the classifiers for each surface type. At the last, you can concatenate all the dataframe and can train the model on the whole data at once and see whether land surface classification increase model performance or not.

## Here is the Demo code that shows you How I trained xtreme gradient boosting decision tree for the phase prediction

In [ ]:
# Now its time to train the model
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
X = df_tran[['10v_BT', '10h_BT', '18v_BT', '18h_BT', '23v_BT', '36v_BT', '36h_BT', '89v_BT', '89h_BT', '166v_BT', '166h_BT', '183-3_BT', '183-7_BT']]
y = df_tran['precip_flag']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42)



from sklearn.naive_bayes import GaussianNB
#from sklearn.naive_bayes import MultinomialNB

# Model for Gaussian Naive Bayes
model = GaussianNB()
# model = MultinomialNB()

# Train the model
model.fit(X_train, y_train)

# Make predictions on the test set
y_pred = model.predict(X_test)

In [ ]:
from sklearn.metrics import confusion_matrix

# Create a confusion matrix
conf_matrix = confusion_matrix(y_test, y_pred)

# Extract TP, TN, FP, FN for labels 1 and 2
TP_1 = conf_matrix[1, 1]  # True Positive for label 1
TN_1 = conf_matrix[0, 0] + conf_matrix[0, 2] + conf_matrix[2, 0] + conf_matrix[2, 2]  # True Negative for label 1
FP_1 = conf_matrix[0, 1] + conf_matrix[2, 1]  # False Positive for label 1
FN_1 = conf_matrix[1, 0] + conf_matrix[1, 2]  # False Negative for label 1

TP_2 = conf_matrix[2, 2]  # True Positive for label 2
TN_2 = conf_matrix[0, 0] + conf_matrix[0, 1] + conf_matrix[1, 0] + conf_matrix[1, 1]  # True Negative for label 2
FP_2 = conf_matrix[0, 2] + conf_matrix[1, 2]  # False Positive for label 2
FN_2 = conf_matrix[2, 0] + conf_matrix[2, 1]  # False Negative for label 2

# Calculate True Positive Rate (TPR) and False Positive Rate (FPR) for labels 1 and 2
TPR_1 = TP_1 / (TP_1 + FN_1)
FPR_1 = FP_1 / (FP_1 + TN_1)

TPR_2 = TP_2 / (TP_2 + FN_2)
FPR_2 = FP_2 / (FP_2 + TN_2)

# Print the results
print(f"Label 1 - TPR: {TPR_1}, FPR: {FPR_1}")
print(f"Label 2 - TPR: {TPR_2}, FPR: {FPR_2}")

Label 1 - TPR: 0.6434240362811792, FPR: 0.18556701030927836
Label 2 - TPR: 0.7231833910034602, FPR: 0.11055694098088113


dic_snow for MultinomialNB:

Label 1 - TPR: 0.8668098818474759, FPR: 0.1104535260403688

Label 2 - TPR: 0.3602860286028603, FPR: 0.23118113479559263


dic_land:

Label 1 - TPR: 0.5839500798142505, FPR: 0.20408668730650154

Label 2 - TPR: 0.0, FPR: 0.000349088878028346


dic_ocean:

Label 1 - TPR: 0.6855644468556444, FPR: 0.09422274793002552

Label 2 - TPR: 0.3393544070914869, FPR: 0.05770107321924336


dic_coast:

Label 1 - TPR: 0.6575963718820862, FPR: 0.31724461105904406

Label 2 - TPR: 0.20761245674740483, FPR: 0.01773344416735938


dic_seaIce:

Label 1 - TPR: 0.632, FPR: 0.21384997740623587

Label 2 - TPR: 0.6124273580688422, FPR: 0.45246326707000867

dic_snow for Guassian NB:

Label 1 - TPR: 0.8700322234156821, FPR: 0.22906802890605532

Label 2 - TPR: 0.6886688668866887, FPR: 0.3340541653794666


dic_land:

Label 1 - TPR: 0.5431722536641996, FPR: 0.11640866873065016

Label 2 - TPR: 0.8755832037325039, FPR: 0.13027996928017874


dic_ocean:

Label 1 - TPR: 0.6843889868438898, FPR: 0.1015065678889373

Label 2 - TPR: 0.9315939120254223, FPR: 0.15501595325902293


dic_coast:

Label 1 - TPR: 0.6434240362811792, FPR: 0.18556701030927836

Label 2 - TPR: 0.7231833910034602, FPR: 0.11055694098088113

dic_seaIce:

Label 1 - TPR: 0.604, FPR: 0.07128332580207862

Label 2 - TPR: 0.8028609745194457, FPR: 0.41248919619706137